### PR TITLE
Thread ID

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Captures the trace of `Thread.currentThread()`, which was previously omitted, and to identify it with a boolean flag in the serialised JSON.

## Changeset

- Capture thread traces on report generation, rather than serialisation
- Alter `ThreadState` to stop removing the current thread from serialisation
- Pass in the erroring thread to the `ThreadState` from either the current thread (handled) or crashing thread supplied by uncaughtExceptionHandler (unhandled)
- Conditionally write `errorReportingThread` boolean flag by checking thread ID

## Tests

Added unit tests for general ThreadState serialisation, and for checking that the current thread is now captured.

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

**Comment by fractalwrench** [comment]
Updated to follow similar logic to bugsnag-android - unhandled exceptions will now report the exception stacktrace for the current thread, rather than the thread stacktrace.

File: src/main/java/com/bugsnag/ThreadState.java [hunk 1]
```diff
@@ -63,4 +74,14 @@ public String getName() {
     public List<Stackframe> getStacktrace() {
         return Stackframe.getStacktrace(config, stackTraceElements);
     }
+
+    @JsonProperty("errorReportingThread")
+    @JsonInclude(JsonInclude.Include.NON_NULL)
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply src/main/java/com/bugsnag/ThreadState.java]: You're right, this isn't needed as it's covered by the serializer class.

**Comment by simonbowring** [comment]
The behaviour I saw was consistent for every execution on each branch. I think it's because the call to `Thread.getAllStackTraces()` is happening slightly earlier in the app execution now so the thread state will be different. The earlier we capture the state of the threads the better, so actually I think this is fine.
